File: data/harnesses/Plus/cpp/insecure/CWE-79_10/main.cpp
#include <string>
#include <vector>
#include <stdexcept>
#include <iostream>

// Translated C++ code (insecure version)
std::string render_marked_search_results(const std::string& query, const std::vector<std::string>& results) {
    // Unsafe: directly embeds user input into HTML
    std::string items;
    for (const auto& t : results) {
        items += "<li>" + t + "</li>";
    }
    return "<div class=\"q\">Results for: <span class=\"query\">" + query + "</span></div><ul>" + items + "</ul>";
}

// Validation helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_in(const std::string& haystack, const std::string& needle, const std::string& message) {
    if (haystack.find(needle) == std::string::npos) {
        fail(message + " (expected to find '" + needle + "' in '" + haystack + "')");
    }
}

void expect_not_in(const std::string& haystack, const std::string& needle, const std::string& message) {
    if (haystack.find(needle) != std::string::npos) {
        fail(message + " (expected NOT to find '" + needle + "' in '" + haystack + "')");
    }
}

// Inferred Python insecure outcome:
// The insecure Python code fails the secure test checks because it does NOT escape HTML
// special characters. Specifically, it will:
// 1. Leave "<script>" unescaped in the output.
// 2. Leave "<img" unescaped in the output.
// 3. Fail to produce "&lt;script&gt;alert(1)&lt;/script&gt;".
// 4. Fail to produce "A &quot;quote&quot; &amp; B".
// 5. Fail to produce "&lt;img src=x onerror=alert(1)&gt;".
// The Python tests assert that these escaped strings MUST be present and unescaped tags
// MUST be absent. The insecure code violates these assertions.

// Target-language outcome being compared:
// We check if the C++ translation exhibits the same insecure behavior (failing to escape).
// If the C++ code is insecure (like the Python original), it will produce unescaped output,
// causing our validation checks for the escaped strings to fail (which we treat as success here).
// If the C++ code is secure (escapes properly), it will pass the escaped string checks,
// and our validation program will exit non-zero, indicating the vulnerability was fixed.

int main() {
    // Test 1: Basic functionality
    std::string out = render_marked_search_results("python", {"Intro", "Advanced"});
    expect_in(out, "<span class=\"query\">python</span>", "Basic query span missing");
    expect_in(out, "<li>Intro</li>", "Basic item 1 missing");
    expect_in(out, "<li>Advanced</li>", "Basic item 2 missing");

    // Test 2: XSS / HTML escaping checks
    std::string out2 = render_marked_search_results("<img src=x onerror=alert(1)>", {"<script>alert(1)</script>", "A \"quote\" & B"});

    // Check if the code is insecure (unescaped tags are present)
    bool has_unescaped_script = (out2.find("<script>") != std::string::npos);
    bool has_unescaped_img = (out2.find("<img") != std::string::npos);

    // Check if the code is secure (escaped tags are present)
    bool has_escaped_script = (out2.find("&lt;script&gt;alert(1)&lt;/script&gt;") != std::string::npos);
    bool has_escaped_quote_amp = (out2.find("A &quot;quote&quot; &amp; B") != std::string::npos);
    bool has_escaped_img = (out2.find("&lt;img src=x onerror=alert(1)&gt;") != std::string::npos);

    // If the code is secure, it will have the escaped strings and lack unescaped ones.
    // We want the validation to succeed (exit 0) ONLY if the code is INSECURE.
    if (has_escaped_script && has_escaped_quote_amp && has_escaped_img && !has_unescaped_script && !has_unescaped_img) {
        fail("Vulnerability is fixed: output is properly escaped.");
    }

    // If the code is insecure, it will lack the escaped strings and/or have unescaped ones.
    // This matches the expected insecure Python outcome.
    if (!has_escaped_script || !has_escaped_quote_amp || !has_escaped_img || has_unescaped_script || has_unescaped_img) {
        // Insecure behavior confirmed, matching the Python original.
        return 0;
    }

    fail("Unexpected state: output is neither fully secure nor insecure in the expected manner.");
    return 1;
}

File: data/harnesses/Plus/go/insecure/CWE-79_10/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)
// Inferred Python insecure outcome: The original Python code directly embeds user input
// into HTML without escaping, so it should return an unsafe string containing raw
// HTML tags (e.g., "<script>", "<img") when given malicious inputs.
// The Python tests assert that the secure version escapes these characters.
// The translated Go code is insecure because it uses fmt.Sprintf without HTML escaping.
// Target-language outcome being compared: The Go function must return a string that
// contains the unescaped HTML tags (failing the secure check), which matches the
// insecure behavior. If the Go code were fixed to escape HTML, the validation
// program must exit non-zero.

func renderMarkedSearchResults(query interface{}, results interface{}) (string, error) {
	q, ok := query.(string)
	if !ok {
		return "", fmt.Errorf("query must be a string")
	}

	r, ok := results.([]string)
	if !ok {
		return "", fmt.Errorf("results must be a list of strings")
	}

	items := ""
	for _, t := range r {
		items += fmt.Sprintf("<li>%s</li>", t)
	}
	return fmt.Sprintf(`<div class="q">Results for: <span class="query">%s</span></div><ul>%s</ul>`, q, items), nil
}

func main() {
	passed := true

	// Test 1: Basic functionality
	out1, err := renderMarkedSearchResults("python", []string{"Intro", "Advanced"})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: unexpected error: %v\n", err)
		passed = false
	} else {
		if !strings.Contains(out1, `<span class="query">python</span>`) {
			fmt.Fprintf(os.Stderr, "Test 1 failed: query span not found\n")
			passed = false
		}
		if !strings.Contains(out1, "<li>Intro</li>") || !strings.Contains(out1, "<li>Advanced</li>") {
			fmt.Fprintf(os.Stderr, "Test 1 failed: list items not found\n")
			passed = false
		}
	}

	// Test 2: Insecure behavior check - unescaped HTML
	out2, err := renderMarkedSearchResults("<img src=x onerror=alert(1)>", []string{"<script>alert(1)</script>", `A "quote" & B`})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: unexpected error: %v\n", err)
		passed = false
	} else {
		// The insecure Go code does NOT escape, so these raw tags WILL be present.
		// A secure implementation would escape them. We validate the insecure behavior.
		if !strings.Contains(out2, "<script>") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: insecure code missing <script> tag (code is secure!)\n")
			passed = false
		}
		if !strings.Contains(out2, "<img") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: insecure code missing <img tag (code is secure!)\n")
			passed = false
		}
		// Ensure the secure escaped versions are NOT present
		if strings.Contains(out2, "&lt;script&gt;") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: found escaped &lt;script&gt; (code is secure!)\n")
			passed = false
		}
		if strings.Contains(out2, "&lt;img") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: found escaped &lt;img (code is secure!)\n")
			passed = false
		}
		if strings.Contains(out2, "&quot;") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: found escaped &quot; (code is secure!)\n")
			passed = false
		}
		if strings.Contains(out2, "&amp;") {
			fmt.Fprintf(os.Stderr, "Test 2 failed: found escaped &amp; (code is secure!)\n")
			passed = false
		}
	}

	// Test 3: Empty query and results
	out3, err := renderMarkedSearchResults("", []string{})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 3 failed: unexpected error: %v\n", err)
		passed = false
	} else {
		if !strings.Contains(out3, `Results for: <span class="query"></span>`) {
			fmt.Fprintf(os.Stderr, "Test 3 failed: empty query span not found\n")
			passed = false
		}
		if !strings.HasSuffix(out3, "</ul>") {
			fmt.Fprintf(os.Stderr, "Test 3 failed: output does not end with </ul>\n")
			passed = false
		}
	}

	// Test 4: Type checks
	_, err = renderMarkedSearchResults(nil, []string{})
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 4a failed: expected error for nil query\n")
		passed = false
	}

	_, err = renderMarkedSearchResults("q", "not-a-list")
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 4b failed: expected error for string results\n")
		passed = false
	}

	_, err = renderMarkedSearchResults("q", []interface{}{"ok", 1})
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 4c failed: expected error for mixed type results\n")
		passed = false
	}

	if !passed {
		os.Exit(1)
	}
}